src/storage/StorageManager.ts

```diff
@@ -0,0 +1,264 @@
+import { BigNumberish, BytesLike } from "ethers";
+import { IStorageModule } from "./interfaces/IStorageManager";
+import {
+    StateSnapshotStruct,
+    BlockConfirmationStruct,
+    ExitChannelBlockStruct,
+    JoinChannelBlockStruct,
+    SignedBlockStruct,
+    BlockStruct
+} from "@typechain-types/contracts/V1/DataTypes";
+import { BlockStorageModule } from "./BlockStorage";
+import { JoinChannelStorageModule } from "./JoinChannelStorage";
+import { ExitChannelStorageModule } from "./ExitChannelStorage";
+import { StateSnapshotStorageModule } from "./StateSnapshotStorage";
+import { BlockHash, StateSnapshotHash } from "@/types/storage";
+import { Codec } from "@/utils";
+import { Type } from "@/utils";
+
+export class StorageModule implements IStorageModule {
+    //#region Global variables
+    private blockStorageModule: BlockStorageModule = new BlockStorageModule();
+    private joinChannelStorageModule: JoinChannelStorageModule =
+        new JoinChannelStorageModule();
+    private exitChannelStorageModule: ExitChannelStorageModule =
+        new ExitChannelStorageModule();
+    private stateSnapshotStorageModule: StateSnapshotStorageModule =
+        new StateSnapshotStorageModule();
+
+    private cachedOnChainStateSnapshot:
+        | {
+              stateSnapshot: StateSnapshotStruct;
+              timestamp: number;
+          }
+        | undefined;
+
+    private latestSignedBlock: SignedBlockStruct | undefined;
+
+    //#endregion
+
+    //#region BlockStorageModule
+
+    insertBlock(signedBlock: SignedBlockStruct): void;
+    insertBlock(
+        signedBlock: SignedBlockStruct,
+        blockHash: BlockHash,
+        fork: number,
+        height: number
+    ): void;
+    insertBlock(blockConfirmation: BlockConfirmationStruct): void;
+    insertBlock(
+        blockConfirmation: BlockConfirmationStruct,
+        blockHash: BlockHash,
+        fork: number,
+        height: number
+    ): void;
+    insertBlock(
+        blockData: SignedBlockStruct | BlockConfirmationStruct,
+        blockHash?: BlockHash,
+        fork?: number,
+        height?: number
+    ): void {
+        if ("signedBlock" in blockData && "signatures" in blockData) {
+            if (blockHash && fork && height) {
+                this.blockStorageModule.insertBlock(
+                    blockData,
+                    blockHash,
+                    fork,
+                    height
+                );
+            } else {
+                this.blockStorageModule.insertBlock(blockData);
+            }
+        } else {
+            if (blockHash && fork && height) {
+                this.blockStorageModule.insertBlock(
+                    blockData,
+                    blockHash,
+                    fork,
+                    height
+                );
+            } else {
+                this.blockStorageModule.insertBlock(blockData);
+            }
+        }
```

Comment:
the external `if/else` is not needed. both arms do the same

```suggestion
            if (blockHash && fork && height) {
                this.blockStorageModule.insertBlock(
                    blockData,
                    blockHash,
                    fork,
                    height
                );
            } else {
                this.blockStorageModule.insertBlock(blockData);
            }
 ```

Reply:
yeah i will remove the nested if but coding it this way is not possible tho since we need to check blockData's type

Reply:
for this one i dont think i can remove the first nested if unless i add some redundency to the function

Reply:
nvm i keep forgetting that ternary operator exist since it doesnt in rust xD